src/shared/telemetry/telemetryEvent.ts

```diff
@@ -5,39 +5,58 @@
 
 'use strict'
 
-import { MetadataEntry, MetricDatum, Unit } from './clienttelemetry'
+import { AwsContext } from '../awsContext'
+import { MetadataEntry, MetricDatum } from './clienttelemetry'
+import { Datum, TelemetryNamespace } from './telemetryTypes'
 
 const NAME_ILLEGAL_CHARS_REGEX = new RegExp('[^\\w+-.:]', 'g')
 const REMOVE_UNDERSCORES_REGEX = new RegExp('_', 'g')
 
-export interface Datum {
-    name: string
-    value: number
-    unit?: Unit
-    metadata?: Map<string, string>
-}
-
 export interface TelemetryEvent {
     namespace: string
     createTime: Date
     data?: Datum[]
 }
 
-export function toMetricData(array: TelemetryEvent[]): MetricDatum[] {
+export function toMetricData(
+    array: TelemetryEvent[],
+    trimmedAwsContext: Pick<AwsContext, 'getCredentialAccountId'>
+): MetricDatum[] {
     return ([] as MetricDatum[]).concat(
         ...array.map( metricEvent => {
 
             const namespace = metricEvent.namespace.replace(REMOVE_UNDERSCORES_REGEX, '')
+            let metadata: MetadataEntry[] = []
+            const account = trimmedAwsContext.getCredentialAccountId()
+
+            if (namespace === TelemetryNamespace.Session) {
+                // this matches JetBrains' functionality: the AWS account ID is not set on session start.
+                metadata.push({
+                    Key: 'account',
+                    Value: undefined
+                })
+            } else {
+                if (account) {
+                    metadata.push({
+                        Key: 'account',
+                        Value: account
+                    })
+                } else {
+                    metadata.push({
+                        Key: 'account',
+                        Value: ''
+                    })
+                }
+            }
 
             if (metricEvent.data !== undefined) {
                 const mappedEventData = metricEvent.data.map( datum => {
-                    let metadata: MetadataEntry[] | undefined
                     let unit = datum.unit
 
                     if (datum.metadata !== undefined) {
-                        metadata = Array.from(datum.metadata).map(entry => {
+                        metadata = metadata.concat(Array.from(datum.metadata).map(entry => {
```

Comment:
1. Consider using `metadata.push(...Array.from(datum.metadata).map(/* ... */))` instead of `metadata = metadata.concat(/* ... */)`. This will allow you to make `metadata` a `const` rather than `let`. It might also have better perf, depending on how `concat` is implemented.

2. Note to self: Add a synchronous generator version of `map` to `shared/utilities/collectionUtils.ts`, which will do the same thing, but but without the overhead of copying to an intermediate array before starting.

Reply:
@mpiroc Looks like #1 isn't working in practice.

Reply:
✔️ Was running into issues with initializing `metadata` outside the lowest `.map`. Moved it in.